File: app/src/main/kotlin/com/softyorch/stroopoverload/game/IncongruenceEngine.kt
package com.softyorch.stroopoverload.game

import com.softyorch.stroopoverload.core.StroopColor
import com.softyorch.stroopoverload.domain.StroopStimulus
import kotlin.random.Random

class IncongruenceEngine(private val rng: Random = Random.Default) {

    fun generate(difficultyLevel: Int): StroopStimulus {
        val inkColor = pick(exclude = emptyList())
        val wordLabel = pick(exclude = listOf(inkColor))
        val audioColor = if (difficultyLevel >= 2) pick(exclude = listOf(inkColor, wordLabel)) else null
        val bgDistractor = if (difficultyLevel >= 3) pick(exclude = listOf(inkColor, wordLabel)) else null

        return StroopStimulus(
            wordLabel = wordLabel,
            inkColor = inkColor,
            audioColor = audioColor,
            bgDistractor = bgDistractor,
        )
    }

    private fun pick(exclude: List<StroopColor>): StroopColor {
        val pool = StroopColor.entries.filter { it !in exclude }
        return if (pool.isEmpty()) StroopColor.entries[rng.nextInt(StroopColor.entries.size)]
        else pool[rng.nextInt(pool.size)]
    }
}

File: lib/game/services/incongruence_engine.dart
import 'dart:math';
import 'package:stroop_overload/core/stroop_color.dart';
import 'package:stroop_overload/domain/stroop_stimulus.dart';

class IncongruenceEngine {
  IncongruenceEngine({Random? rng}) : _rng = rng ?? Random();

  final Random _rng;

  StroopStimulus generate({required int difficultyLevel}) {
    final inkColor = _pick(exclude: []);
    final wordLabel = _pick(exclude: [inkColor]);

    final audioColor =
        difficultyLevel >= 2 ? _pick(exclude: [inkColor, wordLabel]) : null;

    final bgDistractor =
        difficultyLevel >= 3 ? _pick(exclude: [inkColor, wordLabel]) : null;

    return StroopStimulus(
      inkColor: inkColor,
      wordLabel: wordLabel,
      audioColor: audioColor,
      bgDistractor: bgDistractor,
    );
  }

  StroopColor _pick({required List<StroopColor> exclude}) {
    final pool =
        StroopColor.values.where((c) => !exclude.contains(c)).toList();
    if (pool.isEmpty) return StroopColor.values[_rng.nextInt(4)];
    return pool[_rng.nextInt(pool.length)];
  }
}
